### PR TITLE
Lightstep includes and gold

### PR DESCRIPTION
**Pull request by jmillikin-stripe**
### Use only lightstep headers installed to its prefix.

This is another attempt to prevent build failures when using non-sandbox
mode. Instead of trying to find things under `src/c++11/` second, we just
drop that entire directory from the search path.

### Try to use `gold` for linking when using G++ in ASAN mode.

This fixes `/usr/bin/ld: unrecognized option '--push-state'` errors
caused when older versions of G++ use gold-specific options to GNU LD.

### CODE REVIEW

File: bazel/genrule_repository.bzl [hunk 1]
```diff
@@ -84,12 +84,24 @@ def _absolute_bin(path):
...
+
   # Bare minimum cflags to get included test binaries to link.
   #
   # See //tools:bazel.rc for the full set.
   asan_flags = ["-fsanitize=address,undefined"]
   tsan_flags = ["-fsanitize=thread"]
 
+  # Required when using the sanitizer modes in older GCC versions. This is
+  # not a great solution because it doesn't detect GCC when Bazel has wrapped
+  # it in an intermediate script, but it works well enough to keep CI running.
```
Comment (htuch) [inline bazel/genrule_repository.bzl]: Can you describe the versions involved? We can't bump CI to a newer version and set that as the floor?

File: bazel/external/lightstep.BUILD [hunk 2]
```diff
@@ -1,25 +1,21 @@
 load("@envoy//bazel:genrule_repository.bzl", "genrule_cc_deps", "genrule_environment")
 load(":genrule_cmd.bzl", "genrule_cmd")
 
+_HDRS = glob([
+    "src/c++11/lightstep/**/*.h",
+    "src/c++11/mapbox_variant/**/*.hpp",
+])
+
+_PREFIX_HDRS = [hdr.replace("src/c++11/", "_prefix/include/") for hdr in _HDRS] + [
```
Comment (htuch) [inline bazel/external/lightstep.BUILD]: I guess this kind of hackery is worth it for the globs, but this goes back to my original rationale for preferring a `repository_rule` for the full build over `genrule`; you can know all the final headers and glob them in place.

This comment is really non-actionable, but I think as we move more stuff to be `genrule` based, we should look to see what happens with this.

File: bazel/genrule_repository.bzl (hunk 1 above)
Comment (jmillikin-stripe) [reply bazel/genrule_repository.bzl]: @mattklein123 reported errors with `/usr/bin/g++-5`, and the linked StackOverflow post indicates it affects some unknown version range in (5.x, 6.x).

Upgrading CI GCC to v7 would probably fix this but I don't want to start doing wild stuff until the build is more settled.

Comment (htuch) [reply bazel/genrule_repository.bzl]: OK, please just add to comment. Thanks.

Comment (jmillikin-stripe) [reply bazel/genrule_repository.bzl]: Done.

**Pull request merged**